pco/src/float_quant_utils.rs

```diff
@@ -54,48 +56,46 @@ pub(crate) fn split_latents<F: FloatLike>(page_nums: &[F], k: Bitlen) -> Vec<Vec
   vec![primary, secondary]
 }
 
-pub(crate) fn compute_bid_w_sample<F: FloatLike>(sample: &[F]) -> Bid<F::L> {
-  let k = estimate_best_k(sample);
+pub(crate) fn compute_bid<F: FloatLike>(sample: &[F]) -> Option<Bid<F>> {
+  let (k, freq) = estimate_best_k_and_freq(sample);
   // Nothing fancy, we simply estimate that quantizing by k bits results in saving k bits per
-  // number.  This is based on the assumption that FloatQuant will usually be used on datasets that
-  // are exactly quantized.
+  // number, whenever possible. This is based on the assumption that FloatQuant
+  // will usually be used on datasets that are exactly quantized.
+  let bits_saved_per_num = (k as f64) * freq;
   if (k as f64) > QUANT_REQUIRED_BITS_SAVED_PER_NUM {
-    Bid::Candidate {
+    Some(Bid {
       mode: Mode::FloatQuant(k),
-      bits_saved_per_num: k as f64,
-    }
+      bits_saved_per_num,
+      split_fn: Box::new(move |nums| split_latents(nums, k)),
+    })
   } else {
-    Bid::Forfeit
+    None
   }
 }
 
 #[inline(never)]
-pub(crate) fn estimate_best_k<F: FloatLike>(sample: &[F]) -> Bitlen {
-  let thresh = (0.9 * sample.len() as f32).floor() as usize;
-  let mut hist = vec![0; F::PRECISION_BITS.try_into().unwrap()];
-  for num_tz in sample.iter().map(|&x| {
-    cmp::min(
-      F::PRECISION_BITS,
-      // Using the fact that significand bits come last in
-      // the floating-point representations we care about
-      x.trailing_zeros(),
-    )
-  }) {
-    hist[num_tz as usize] += 1
+pub(crate) fn estimate_best_k_and_freq<F: FloatLike>(sample: &[F]) -> (Bitlen, f64) {
+  let thresh = (REQUIRED_QUANTIZED_PROPORTION * sample.len() as f64) as usize;
+  let mut hist = vec![0; (F::PRECISION_BITS + 1) as usize];
+  for x in sample {
+    // Using the fact that significand bits come last in
+    // the floating-point representations we care about
+    let trailing_mantissa_zeros = cmp::min(F::PRECISION_BITS, x.trailing_zeros());
+    hist[trailing_mantissa_zeros as usize] += 1
   }
-  hist
-    .iter()
-    .enumerate()
-    .rev()
-    .scan(0_usize, |csum, (i, x)| {
-      if *csum >= thresh {
-        return None;
-      }
-      *csum = *csum + x;
-      Some(i)
-    })
-    .last()
-    .expect("nums is nonempty") as Bitlen
+
+  let mut rev_csum = 0;
+  for (k, &occurrences) in hist.iter().enumerate().rev() {
+    rev_csum += occurrences;
+    if rev_csum >= thresh {
+      return (
+        k as Bitlen,
+        rev_csum as f64 / sample.len() as f64,
+      );
+    }
+  }
+
+  unreachable!("nums should be nonempty")
```

Comment:
I'm surprised so little revision was needed, since I hadn't run the code in a unit test yet lol.

For my own edification, do you recall what went wrong in the first draft?

Reply:
Sure!

* the hist went from `[0, PRECISION_BITS)`, but we could have trailing zeros in `[0, PRECISION_BITS]`, so we need the + 1
* the csum omitted the current element (addition should come before the thresh check)
* I think you would have wanted to take `.first()` instead of `.last()`. For me it's a little easier to reason about as a loop that terminates as soon as it hits the condition.

Reply:
Ah that's true, I initialized `hist` too small, and this would have resulted in an out-of-bounds assignment if any of the `x`s had all their mantissa bits equal to zero.

The csum lagging by one element was intentional, and taking `last()` was also intentional -- we are iterating over all $r$ such that
$$\big| \lbrace x \ |\ \text{TZ}(x) \geq r \rbrace \big| < \texttt{thresh}$$
and we want one before the smallest such $r$.

I'm fine with the rewrite though.  You're probably right that reasoning about the `rev` of an `enumerate` is more cognitive load for most readers than reasoning about an ascending index that gets subtracted from stuff.

Reply:
oh, the rewrite also uses the `rev` of an `enum`.  oh well lol, still fine